README.md
# ApiHammer

[![Build Status](https://travis-ci.org/notEthan/api_hammer.svg?branch=master)](https://travis-ci.org/notEthan/api_hammer)

a collection of little tools I have used when creating APIs. these are generally too small to have their own 
library, so have been collected here. 

## ApiHammer::Rails

A module for inclusion in a Rails application controller inheriting from ActionController::Base (Rocketpants 
should also work, but no guarantees). 

### #halt

You know that pattern in rails, `render(some_stuff) and return`? 

Don't use it. It's wrong. `and` is used to short-circuit, but you don't actually want your `return` to be conditional on the `render`. Since it is conditional on the return value of `render`, if the return value of `render` is false-ish (`false` or `nil`), you may end end up failing to return. The actual return value of `render` is undocumented. If you follow all the control paths that can affect the return value, last time I looked there were 37 different paths. None of them returns false, but if any control path changes to return false then `render and return` will break. 

What you really mean is `render(some_stuff); return`, but semicolons are (rightly) frowned upon 
in ruby. Really you should just put the `return` on the next line. 

Or, use `#halt`, kindly provided by ApiHammer.

`ApiHammer#halt` is based on the Sinatra pattern of `throw(:halt, response)`. It uses exceptions because Rails has baked-in exception rescuing (but not, at least as far as I have found, throw catching), but that is hidden away. 

### #check_required_params

Pass it an parameters which are required, and if they are missing or incorrect it will halt with 422 (not 400 - 400 is wrong). 

## Other

Various other things. This readme is incomplete and will be updated soon. 


CHANGELOG.md
# v0.19.0
- drop support for ruby < 2.2
- count time monotonically
- bugfix getting current logger tags
- prepare for ur

# v0.18.2
- bugfix utf8 http body handling
- report to Rollbar on 4xx if Rollbar available and DEBUG_4XX indicates so

# v0.18.1
- improve supported media types for sinatra
- report to Rollbar on 4xx if Rollbar available

# v0.18.0
- add option to hc for basic auth
- add option to hc to make it consider the body as text
- faraday logger compatibility with old faraday 0.8
- fix ruby warnings

# v0.17.0
- fix ApiHammer::ShowTextExceptions middleware to not rescue certain exceptions
- don't insert or rely on Rack::Accept middleware for sinatra; use Rack::Accept more directly
- improve hc User-Agent

# v0.16.0
- fix dependency on json pure parser, which moved from the json gem to json_pure gem

# v0.15.0
- fix check_required_params for ActionController::Params (really this time?)
- Dir.glob_escape / Pathname#glob_escape
- Kernel#ycomb - the y combinator
- sinatra helper #check_params_and_object_consistent
- fix sinatra #format_response with 204 no content
- improve testing with up-to-date ruby (2.3.1), rack (2), actionpack (5), as well as older ones

# v0.14.1
- task cucumber:trailing_whitespace to clean trailing whitespace from cucumber features #54

# v0.14.0
- some rails 5 support
  - check_required_params to support ActionController::Parameters #44
  - handle not loading deprecated log_tailer when not found #33
  - use ruby String#bytesize instead of Rack::Util #34
- fix same bug as v0.13.3 with logging non-ascii bodies on faraday logger #35
- remove ActiveRecord cache_find_by
- option :log_bodies for request loggers
- fix warnings when run with -w

# v0.13.3
- fix bug when logging non-ascii bodies with filtration enabled #31

# v0.13.2
- fix with_indifferent_access usage when we don't depend on activesupport #29

# v0.13.1
- ApiHammer::Sinatra class method use_with_lint
- set up Rack::Accept middleware in sinatra as api hammer methods rely on that

# v0.13.0
- ApiHammer::Sinatra, with some useful sinatra methods
  - #halt, #halt_error, #halt_unprocessable_entity and friends
  - a more api-appropriate 404 for unknown routes
  - parsing request bodies in accordance with content-type, with appropriate errors
  - formatting response bodies, minding accept headers, with appropriate errors

# v0.12.0
- hc --input option
- rails 4 support for unmunged_request_params

# v0.11.1
- bugfix trailingnewline

# v0.11.0
- improved handling of text and binary bodies in logging middleware and hc

# v0.10.2
- rails request logging logs exception backtrace

# v0.10.1
- Rack RequestLogger works around rails' exception app path mangling

# v0.10.0
- JsonScriptEscapeHelper

# v0.9.2
- bugfix form encoded filtering

# v0.9.1
- recognize `app.config.api_hammer_request_logging_options` for request logger options

# v0.9.0
- rack request logger logs ids in arrays of hashes when logging ids
- filtered logging of sensitive keys in bodies of requests (json and form encoded)
- logstash filter for oauth headers and oauthenticator log entries

# v0.8.1
- request log format tweaks

# v0.8.0
- log request and response bodies - not just IDs from them - if they aren't too big

# v0.7.1
- logstash filters for sidekiq, activesupport tags, and of course ApiHammer's request logging 
- use i18n for errors and add error_message to response
- hc assumes http if no protocol specified

# v0.6.3
- add request role to the request logging

# v0.6.2
- ApiHammer::RequestLogger response body fix

# v0.6.1
- broken release, yanked

# v0.6.0
- ApiHammer::RailsOrSidekiqLogger

# v0.5.0
- rack request logger logs all request and response headers
- fix id / uuid / guid logging in rack request logger
- faraday request logger does not log binary bodies 

# 0.4.3
- bugfix

# 0.4.2
- bugfix encoding in faraday request logger

# 0.4.1
- bugfix

# 0.4.0
- ApiHammer::Faraday::RequestLogger

# 0.3.3
- be a little lazier about initializing ActiveRecord::Base.finder_cache - only on first actual usage 

# 0.3.2
- ActiveRecord::Base.cache_find_by support finding on associations, fix bind detection with symbols 

# 0.3.1
- bugfix ActiveRecord::Base.cache_find_by

# 0.3.0
- ActiveRecord::Base.cache_find_by

# 0.2.2
- RequestLogger, in addition to logging response bodies on error, logs id/uuid fields from request body and 
  response body if there's no error
- support a logger instead of a device in FaradayOutputter

# 0.2.1
- small RequestLogger tweaks

# 0.2.0
- Rails#unmunged_request_params
- hc --pretty
- hc default User-Agent set
- hc doc improved, middlewares moved lib
- RequestLogger improved
- ApiHammer::RailsRequestLogging railtie
- README improved marginally 

# 0.1.0
- Object#public_instance_exec
- Obect#public_instance_eval
- hc

# 0.0.3
- rake cucumber:pretty_json
- Rails#find_or_halt

# 0.0.2

- Weblink#to_s

# 0.0.1

- rake gem:available_updates
- Rails#halt
- Rails#check_required_params
- Weblink
- RequestLogger
- ShowTextExceptions
- TrailingNewline


lib/api_hammer/halt.rb
# the contents of this file are to let you halt a controller in its processing without having to have a 
# return in the actual action. this lets helper methods which do things like parameter validation halt.
#
# it is desgined to function similarly to Sinatra's handling of throw(:halt), but is based around exceptions 
# because rails doesn't catch anything, just rescues. 

module ApiHammer
  # an exception raised to stop processing an action and render the body given as the 'body' argument 
  # (which is expected to be a JSON-able object)
  class Halt < StandardError
    def initialize(message, body, render_options={})
      super(message)
      @body = body
      @render_options = render_options
    end

    attr_reader :body, :render_options
  end
end

module ApiHammer::Rails
  unless instance_variables.any? { |iv| iv.to_s == '@halt_included' }
    @halt_included = proc do |controller_class|
      controller_class.send(:rescue_from, ApiHammer::Halt, :with => :handle_halt)
    end
    (@on_included ||= Set.new) << @halt_included
  end

  # handle a raised ApiHammer::Halt or subclass and render it
  def handle_halt(halt)
    render_options = halt.render_options ? halt.render_options.dup : {}
    # rocket pants does not have a render method, just render_json 
    if respond_to?(:render_json, true)
      render_json(halt.body || {}, render_options)
    else
      render_options[:json] = halt.body || {}
      render(render_options)
    end
  end

  # halt and render the given body 
  def halt(status, body, render_options = {})
    raise(ApiHammer::Halt.new(body.inspect, body, render_options.merge(:status => status)))
  end

  # halt and render the given errors in the body on the 'errors' key 
  def halt_error(status, errors, options = {})
    errors_as_json = errors.respond_to?(:as_json) ? errors.as_json : errors
    unless errors_as_json.is_a?(Hash)
      raise ArgumentError, "errors be an object representable in JSON as a Hash; got errors = #{errors.inspect}"
    end
    unless errors_as_json.keys.all? { |k| k.is_a?(String) || k.is_a?(Symbol) }
      raise ArgumentError, "errors keys must all be string or symbol; got errors = #{errors.inspect}"
    end
    unless errors_as_json.values.all? { |v| v.is_a?(Array) && v.all? { |e| e.is_a?(String) } }
      raise ArgumentError, "errors values must all be arrays of strings; got errors = #{errors.inspect}"
    end
    render_options = options.dup.with_indifferent_access
    body = {'errors' => errors}
    error_message = render_options.delete('error_message') || begin
      error_values = errors.values.inject([], &:+)
      if error_values.size <= 1
        error_values.first
      else
        # sentencify with periods 
        error_values.map { |v| v =~ /\.\s*\z/ ? v : v + '.' }.join(' ')
      end
    end
    body['error_message'] = error_message if error_message
    halt(status, body, render_options)
  end

  # attempts to find and return the given model (an ActiveRecord::Base subclass) with the given attributes. 
  # halts with 404 (does not return) if that fails. options[:status] may specify a different status if that 
  # is required.
  #
  # e.g.:
  #
  #     find_or_halt(User, :email => 'user@example.com')
  #
  def find_or_halt(model, find_attrs, options={})
    options = {:status => 404}.merge(options)
    record = model.where(find_attrs).first
    unless record
      attributes = find_attrs.map{|attr, val| "#{attr}: #{val}" }.join(", ")
      model_name = model.table_name
      model_name = model_name.singularize if model_name.respond_to?(:singularize)
      message = I18n.t(:"errors.unknown_record_with_attributes", :default => "Unknown %{model_name}! %{attributes}",
        :model_name => model_name,
        :attributes => attributes
      )
      halt_error(options[:status], {model_name => [message]})
    end
    record
  end

  module HaltMethods
    # halt with status 200 OK, responding with the given body object 
    def halt_ok(body, render_options = {})
      halt(200, body, render_options)
    end

    # halt with status 201 Created, responding with the given body object 
    def halt_created(body, render_options = {})
      halt(201, body, render_options)
    end

    # halt with status 202 Accepted, responding with the given body object 
    def halt_accepted(body, render_options = {})
      halt(202, body, render_options)
    end

    # halt with status 203 Non-Authoritative Information, responding with the given body object 
    def halt_non_authoritative_information(body, render_options = {})
      halt(203, body, render_options)
    end

    # halt with status 204 No Content
    def halt_no_content(render_options = {})
      halt(204, '', render_options)
    end

    # halt with status 205 Reset Content
    def halt_reset_content(render_options = {})
      halt(205, '', render_options)
    end

    # halt with status 206 Partial Content, responding with the given body object 
    def halt_partial_content(body, render_options = {})
      halt(206, body, render_options)
    end

    # halt with status 207 Multi-Status, responding with the given body object 
    def halt_multi_status(body, render_options = {})
      halt(207, body, render_options)
    end

    # halt with status 208 Already Reported, responding with the given body object 
    def halt_already_reported(body, render_options = {})
      halt(208, body, render_options)
    end

    # halt with status 226 IM Used, responding with the given body object 
    def halt_im_used(body, render_options = {})
      halt(226, body, render_options)
    end

    # halt with status 300 Multiple Choices, responding with the given body object 
    def halt_multiple_choices(body, render_options = {})
      halt(300, body, render_options)
    end

    # halt with status 301 Moved Permanently, responding with the given body object 
    def halt_moved_permanently(body, render_options = {})
      halt(301, body, render_options)
    end

    # halt with status 302 Found, responding with the given body object 
    def halt_found(body, render_options = {})
      halt(302, body, render_options)
    end

    # halt with status 303 See Other, responding with the given body object 
    def halt_see_other(body, render_options = {})
      halt(303, body, render_options)
    end

    # halt with status 304 Not Modified
    def halt_not_modified(render_options = {})
      halt(304, '', render_options)
    end

    # halt with status 305 Use Proxy, responding with the given body object 
    def halt_use_proxy(body, render_options = {})
      halt(305, body, render_options)
    end

    # halt with status 307 Temporary Redirect, responding with the given body object 
    def halt_temporary_redirect(body, render_options = {})
      halt(307, body, render_options)
    end

    # halt with status 308 Permanent Redirect, responding with the given body object 
    def halt_permanent_redirect(body, render_options = {})
      halt(308, body, render_options)
    end

    # halt with status 400 Bad Request, responding with the given errors object on the 'errors' key
    def halt_bad_request(errors, render_options = {})
      halt_error(400, errors, render_options)
    end

    # halt with status 401 Unauthorized, responding with the given errors object on the 'errors' key
    def halt_unauthorized(errors, render_options = {})
      halt_error(401, errors, render_options)
    end

    # halt with status 402 Payment Required, responding with the given errors object on the 'errors' key
    def halt_payment_required(errors, render_options = {})
      halt_error(402, errors, render_options)
    end

    # halt with status 403 Forbidden, responding with the given errors object on the 'errors' key
    def halt_forbidden(errors, render_options = {})
      halt_error(403, errors, render_options)
    end

    # halt with status 404 Not Found, responding with the given errors object on the 'errors' key
    def halt_not_found(errors, render_options = {})
      halt_error(404, errors, render_options)
    end

    # halt with status 405 Method Not Allowed, responding with the given errors object on the 'errors' key
    def halt_method_not_allowed(errors, render_options = {})
      halt_error(405, errors, render_options)
    end

    # halt with status 406 Not Acceptable, responding with the given errors object on the 'errors' key
    def halt_not_acceptable(errors, render_options = {})
      halt_error(406, errors, render_options)
    end

    # halt with status 407 Proxy Authentication Required, responding with the given errors object on the 'errors' key
    def halt_proxy_authentication_required(errors, render_options = {})
      halt_error(407, errors, render_options)
    end

    # halt with status 408 Request Timeout, responding with the given errors object on the 'errors' key
    def halt_request_timeout(errors, render_options = {})
      halt_error(408, errors, render_options)
    end

    # halt with status 409 Conflict, responding with the given errors object on the 'errors' key
    def halt_conflict(errors, render_options = {})
      halt_error(409, errors, render_options)
    end

    # halt with status 410 Gone, responding with the given errors object on the 'errors' key
    def halt_gone(errors, render_options = {})
      halt_error(410, errors, render_options)
    end

    # halt with status 411 Length Required, responding with the given errors object on the 'errors' key
    def halt_length_required(errors, render_options = {})
      halt_error(411, errors, render_options)
    end

    # halt with status 412 Precondition Failed, responding with the given errors object on the 'errors' key
    def halt_precondition_failed(errors, render_options = {})
      halt_error(412, errors, render_options)
    end

    # halt with status 413 Request Entity Too Large, responding with the given errors object on the 'errors' key
    def halt_request_entity_too_large(errors, render_options = {})
      halt_error(413, errors, render_options)
    end

    # halt with status 414 Request-URI Too Long, responding with the given errors object on the 'errors' key
    def halt_request_uri_too_long(errors, render_options = {})
      halt_error(414, errors, render_options)
    end

    # halt with status 415 Unsupported Media Type, responding with the given errors object on the 'errors' key
    def halt_unsupported_media_type(errors, render_options = {})
      halt_error(415, errors, render_options)
    end

    # halt with status 416 Requested Range Not Satisfiable, responding with the given errors object on the 'errors' key
    def halt_requested_range_not_satisfiable(errors, render_options = {})
      halt_error(416, errors, render_options)
    end

    # halt with status 417 Expectation Failed, responding with the given errors object on the 'errors' key
    def halt_expectation_failed(errors, render_options = {})
      halt_error(417, errors, render_options)
    end

    # halt with status 418 I'm a teapot, responding with the given errors object on the 'errors' key
    def halt_im_a_teapot(errors, render_options = {})
      halt_error(418, errors, render_options)
    end

    # halt with status 419 Authentication Timeout, responding with the given errors object on the 'errors' key
    def halt_authentication_timeout(errors, render_options = {})
      halt_error(419, errors, render_options)
    end

    # halt with status 422 Unprocessable Entity, responding with the given errors object on the 'errors' key
    def halt_unprocessable_entity(errors, render_options = {})
      halt_error(422, errors, render_options)
    end

    # halt with status 423 Locked, responding with the given errors object on the 'errors' key
    def halt_locked(errors, render_options = {})
      halt_error(423, errors, render_options)
    end

    # halt with status 424 Failed Dependency, responding with the given errors object on the 'errors' key
    def halt_failed_dependency(errors, render_options = {})
      halt_error(424, errors, render_options)
    end

    # halt with status 425 Unordered Collection, responding with the given errors object on the 'errors' key
    def halt_unordered_collection(errors, render_options = {})
      halt_error(425, errors, render_options)
    end

    # halt with status 426 Upgrade Required, responding with the given errors object on the 'errors' key
    def halt_upgrade_required(errors, render_options = {})
      halt_error(426, errors, render_options)
    end

    # halt with status 428 Precondition Required, responding with the given errors object on the 'errors' key
    def halt_precondition_required(errors, render_options = {})
      halt_error(428, errors, render_options)
    end

    # halt with status 429 Too Many Requests, responding with the given errors object on the 'errors' key
    def halt_too_many_requests(errors, render_options = {})
      halt_error(429, errors, render_options)
    end

    # halt with status 431 Request Header Fields Too Large, responding with the given errors object on the 'errors' key
    def halt_request_header_fields_too_large(errors, render_options = {})
      halt_error(431, errors, render_options)
    end

    # halt with status 444 No Response, responding with the given errors object on the 'errors' key
    def halt_no_response(errors, render_options = {})
      halt_error(444, errors, render_options)
    end

    # halt with status 451 Unavailable For Legal Reasons, responding with the given errors object on the 'errors' key
    def halt_unavailable_for_legal_reasons(errors, render_options = {})
      halt_error(451, errors, render_options)
    end

    # halt with status 451 Redirect, responding with the given errors object on the 'errors' key
    def halt_redirect(errors, render_options = {})
      halt_error(451, errors, render_options)
    end

    # halt with status 500 Internal Server Error, responding with the given errors object on the 'errors' key
    def halt_internal_server_error(errors, render_options = {})
      halt_error(500, errors, render_options)
    end

    # halt with status 501 Not Implemented, responding with the given errors object on the 'errors' key
    def halt_not_implemented(errors, render_options = {})
      halt_error(501, errors, render_options)
    end

    # halt with status 502 Bad Gateway, responding with the given errors object on the 'errors' key
    def halt_bad_gateway(errors, render_options = {})
      halt_error(502, errors, render_options)
    end

    # halt with status 503 Service Unavailable, responding with the given errors object on the 'errors' key
    def halt_service_unavailable(errors, render_options = {})
      halt_error(503, errors, render_options)
    end

    # halt with status 504 Gateway Timeout, responding with the given errors object on the 'errors' key
    def halt_gateway_timeout(errors, render_options = {})
      halt_error(504, errors, render_options)
    end

    # halt with status 505 HTTP Version Not Supported, responding with the given errors object on the 'errors' key
    def halt_http_version_not_supported(errors, render_options = {})
      halt_error(505, errors, render_options)
    end

    # halt with status 506 Variant Also Negotiates, responding with the given errors object on the 'errors' key
    def halt_variant_also_negotiates(errors, render_options = {})
      halt_error(506, errors, render_options)
    end

    # halt with status 507 Insufficient Storage, responding with the given errors object on the 'errors' key
    def halt_insufficient_storage(errors, render_options = {})
      halt_error(507, errors, render_options)
    end

    # halt with status 508 Loop Detected, responding with the given errors object on the 'errors' key
    def halt_loop_detected(errors, render_options = {})
      halt_error(508, errors, render_options)
    end

    # halt with status 509 Bandwidth Limit Exceeded, responding with the given errors object on the 'errors' key
    def halt_bandwidth_limit_exceeded(errors, render_options = {})
      halt_error(509, errors, render_options)
    end

    # halt with status 510 Not Extended, responding with the given errors object on the 'errors' key
    def halt_not_extended(errors, render_options = {})
      halt_error(510, errors, render_options)
    end

    # halt with status 511 Network Authentication Required, responding with the given errors object on the 'errors' key
    def halt_network_authentication_required(errors, render_options = {})
      halt_error(511, errors, render_options)
    end
=begin
# the above methods are generated with the following script

require 'nokogiri'
require 'faraday'
wpcodes = Nokogiri::HTML(Faraday.get('https://en.wikipedia.org/wiki/List_of_HTTP_status_codes').body)
dts = wpcodes.css('dt')
puts(dts.map do |dt|
  if dt.text =~ /\A\s*(\d+)\s+([a-z0-9 \-']+)/i
    status = $1.to_i
    name = $2.strip
    underscore = name.split(/[\s-]/).map{|word| word.downcase.gsub(/\W/, '') }.join('_')
    if ([100..199, 490..499, 520..599].map(&:to_a).inject([], &:+) + [306, 420, 440, 449, 450]).include?(status)
      # exclude these. 1xx isn't really a thing that makes sense at this level and the others are 
      # nonstandard or particular particular web servers or such 
      nil
    elsif [204, 205, 304].include?(status)
      # ones with no body
%Q(
  # halt with status #{status} #{name}
  def halt_#{underscore}(render_options = {})
    halt(#{status}, '', render_options)
  end
)
    elsif (400..599).include?(status)
      # body goes on an errors object
%Q(
  # halt with status #{status} #{name}, responding with the given errors object on the 'errors' key
  def halt_#{underscore}(errors, render_options = {})
    halt_error(#{status}, errors, render_options)
  end
)
    else
%Q(
  # halt with status #{status} #{name}, responding with the given body object 
  def halt_#{underscore}(body, render_options = {})
    halt(#{status}, body, render_options)
  end
)
    end
  end
end.compact.join)

=end
  end

  include HaltMethods
end
